### PR TITLE
Support restricted PSA for worker pods

### PR DESCRIPTION
**Pull request by mhenriks**
Signed-off-by: Michael Henriksen <mhenriks@redhat.com>

**What this PR does / why we need it**:

Attempt to get in compliance with restricted PSA. See:  https://kubernetes.io/docs/concepts/security/pod-security-admission/

- [X] Remove root pods
- [x] Remove selinux params
- [x] SecurityContextConstraint update
- [x] Run func test in restricted SCC

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Support Restricted PSA for worker pods
```

### CODE REVIEW

**Comment by awels** [comment]
I suspect the nbdkit socket will be an issue here.

**Comment by mhenriks** [comment]
> I suspect the nbdkit socket will be an issue here.

please elaborate

**Comment by awels** [comment]
Basically when I was working on https://github.com/kubevirt/containerized-data-importer/pull/2331 I also tried removing root as part of that, and when I was running tests, they started failing because I was unable to access the nbdkit socket since I wasn't root. I didn't look into the exact reason much more though.

**Comment by awels** [comment]
But it looks like the tests are passing, so maybe not an issue.

File: pkg/controller/clone-controller.go [hunk 1]
```diff
@@ -733,6 +723,7 @@ func MakeCloneSourcePodSpec(sourceVolumeMode corev1.PersistentVolumeMode, image,
 			{
 				Name:      DataVolName,
 				MountPath: common.ClonerMountPath,
+				ReadOnly:  true,
```
Comment (akalenyu) [inline pkg/controller/clone-controller.go:726]: Nice, is there a real impact though? IIUC the other ReadOnly is just about enforcing this one

**Review by akalenyu — COMMENTED** [review]
> Basically when I was working on #2331 I also tried removing root as part of that, and when I was running tests, they started failing because I was unable to access the nbdkit socket since I wasn't root. I didn't look into the exact reason much more though.

Really interested about why last time around we saw imports failing on block without root.
Maybe related to us switching to less nbdkit usages?

> Attempt to get in compliance with restricted PSA. See: https://kubernetes.io/docs/concepts/security/pod-security-admission/

Reading the k8s [docs](https://kubernetes.io/docs/concepts/security/pod-security-standards/) IIUC we need to **explicitly** set more things to comply with "restricted":
- allowPrivilegeEscalation to false (done by #2331 )
- runAsNonRoot (on pod level is easier)
- seccompProfile (pod level)

What am I missing?

> * SecurityContextConstraint update

If we end up changing our version of the SCC we should probably make the reconciliation of it more strict so it gets upgraded properly. Currently it'll only reconcile if https://github.com/kubevirt/containerized-data-importer/blob/13a9b1299ea91219c607ac3d8dbd5ab4a3d4cc89/pkg/operator/controller/scc.go#L105-L109

File: pkg/importer/http-datasource.go [hunk 2]
```diff
@@ -46,8 +46,8 @@ import (
 
 const (
 	tempFile         = "tmpimage"
-	nbdkitPid        = "/var/run/nbdkit.pid"
-	nbdkitSocket     = "/var/run/nbdkit.sock"
+	nbdkitPid        = "/tmp/nbdkit.pid"
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/importer/http-datasource.go:49]: I've never heard of any tmp cleanup happening in the absence of a reboot.

File: pkg/importer/registry-datasource.go [hunk 3]
```diff
@@ -165,7 +165,7 @@ func getImageFileName(dir string) (string, error) {
 }
 
 func createCertificateDir(registryCertDir string) (string, error) {
-	allCerts := "/var/local/all_certs"
+	allCerts := "/tmp/all_certs"
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/importer/registry-datasource.go:168]: non root cannot write to `/var/local`

File: pkg/operator/controller/scc.go [hunk 4]
```diff
@@ -37,6 +37,34 @@ import (
 
 const sccName = "containerized-data-importer"
 
+func setSCC(scc *secv1.SecurityContextConstraints) {
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/operator/controller/scc.go:40]: Cool, maybe next time

**Comment by mhenriks** [comment]
> Awesome We probably won't hit the same issues as in the PR description here, right? [kubevirt/kubevirt#8436](https://github.com/kubevirt/kubevirt/pull/8436)

What issues, specifically?  They are attacking the issue in a totally different way for now

File: pkg/importer/registry-datasource.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (akalenyu) [reply pkg/importer/registry-datasource.go:168]: gotcha

File: pkg/util/util.go [hunk 5]
```diff
@@ -201,27 +201,21 @@ func StreamDataToFile(r io.Reader, fileName string) error {
...
-
-	var tarOptions string
-	var args = arg
-	if len(arg) > 0 {
-		tarOptions = arg[0]
-		args = arg[1:]
-	}
-	options := fmt.Sprintf("-%s%s", tarOptions, "xvC")
-	untar := exec.Command("/usr/bin/tar", "--no-same-owner", options, destDir, strings.Join(args, ""))
+	untar := exec.Command("/usr/bin/tar", "--no-same-owner", "-xvC", destDir)
```
[the remark this answers is not in the document]
Comment (akalenyu) [reply pkg/util/util.go:206]: I saw you took up the suggestion in the clone source pod but not here. Intended?

File: tests/framework/framework.go [hunk 6]
```diff
@@ -218,6 +218,15 @@ func (f *Framework) AfterEach() {
 
 // CreateNamespace instantiates a new namespace object with a unique name and the passed-in label(s).
 func (f *Framework) CreateNamespace(prefix string, labels map[string]string) (*v1.Namespace, error) {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	// pod-security.kubernetes.io/<MODE>: <LEVEL>
+	labels["pod-security.kubernetes.io/enforce"] = "restricted"
+	if utils.IsOpenshift(f.K8sClient) {
+		labels["security.openshift.io/scc.podSecurityLabelSync"] = "false"
```
Comment (akalenyu) [inline tests/framework/framework.go:227]: Nice OpenShift awareness!

**Comment by akalenyu** [comment]
> > Awesome We probably won't hit the same issues as in the PR description here, right? [kubevirt/kubevirt#8436](https://github.com/kubevirt/kubevirt/pull/8436)
> 
> What issues, specifically? They are attacking the issue in a totally different way for now

Looks good just let me know if you want to tackle the few small replies, otherwise I think we are good

**Review by jean-edouard — APPROVED** [review]
Just one minor comment in-line

**Pull request merged**